src/render-queue.js

```diff
@@ -16,6 +16,12 @@ export function rerender() {
 	let p, list = items;
 	items = [];
 	while ( (p = list.pop()) ) {
-		if (p._dirty) renderComponent(p);
+		if (p._dirty) {
+			try {
```

Comment:
I'm worried about the try/catch for each cycle of the while loop. 

Looking at `renderComponent`, it [_already_ has its own try/catch](https://github.com/rpetrich/preact/blob/3a3d7cb1213bdd643e4a35323cc4478065e4e061/src/vdom/component.js#L157) block that we can & should reuse.

Instead, `catchErrorInComponent` should be passed to `renderComponent` somehow (/cc @developit) -- or, better, yet, we import the function into `vdom/component.js` & call it directly from the `catch` phase.

Reply:
Errors in render/child components should finish mounting/updating the component before dispatching `componentDidCatch` and potentially re-rendering. This is the try/catch around render/child components inside `renderComponent` and won't catch errors in the components mounting/update methods. Maybe a single try/catch inside `renderComponent` and booleans representing how much progress into the method has completed is a better approach? (since the error needs to be handled differently depending on where it comes from)

Reply:
Errors in render/child components should finish mounting/updating the component before dispatching `componentDidCatch` and potentially re-rendering. This is the try/catch around render/child components inside `renderComponent` and won't catch errors in the components mounting/update methods. Maybe a single try/catch inside `renderComponent` and booleans representing how much progress into the method has completed is a better approach? (since the error needs to be handled differently depending on where it comes from)

Reply:
Got it 👍 What would the boolean approach look like? Would it be a single key assignment on every component, every render?

Reply:
Boolean approach evolved into a nested try/catch with some additional state tracking on the component itself (see: the new `_caught` property) to properly handle cases where a component fails repeatedly and should bubble its errors to an ancestor.

I also merged the latest master and added tests for `getDerivedStateFromProps` and `getSnapshotBeforeUpdate` to ensure they're properly integrated with `componentDidCatch`